Add index block api (#58094)
elastic/elasticsearch: docs/reference/index-modules/blocks.asciidoc

    allow writes and metadata changes.

`index.blocks.read_only_allow_delete`::

    Similar to `index.blocks.read_only`, but also allows deleting the index to
    make more resources available. The <<disk-based-shard-allocation,disk-based shard
    allocator>> may add and remove this block automatically.

Deleting documents from an index to release resources - rather than deleting the index itself - can increase the index size over time. When `index.blocks.read_only_allow_delete` is set to `true`, deleting documents is not permitted. However, deleting the index itself releases the read-only index block and makes resources available almost immediately.

IMPORTANT: {es} adds and removes the read-only index block automatically when the disk utilization falls below the high watermark, controlled by <<cluster-routing-flood_stage,cluster.routing.allocation.disk.watermark.flood_stage>>.

`index.blocks.read`::

    Set to `true` to disable read operations against the index.

`index.blocks.write`::

    Set to `true` to disable data write operations against the index. Unlike `read_only`,
    this setting does not affect metadata. For instance, you can close an index with a `write`

>     block, but not an index with a `read_only` block.

Reviewer: Tweak for clarity.
```suggestion
    block, but you cannot close an index with a `read_only` block.
```

`index.blocks.metadata`::

    Set to `true` to disable index metadata reads and writes.

[discrete]
[[add-index-block]]
=== Add index block API

> Adds an index block, and ensures that, once successfully returning to the
> user, all shards of the index are properly accounting for the block, for
> example that all in-flight writes to an index have been completed after
> adding the write block.

Reviewer: The second part of this sentence (`, and ensures that...`) confused me. Is it saying `If successful, the API returns a confirmation`?  Either way, I'm not sure it's needed. 

If you decide to keep it, I would reword and split into a separate sentence.
```suggestion
Adds an index block to an index.
```

[source,console]
--------------------------------------------------
PUT /twitter/_block/write
--------------------------------------------------
// TEST[setup:twitter]

> The blocks available are `read_only`, `read_only_allow_delete`, `read`, `write`,
> and `metadata`, and their meaning is described above in the settings section.

Reviewer: I'd move this to the `<block>` path parameter definition below.
```suggestion
```

[discrete]
[[add-index-block-api-request]]
==== {api-request-title}

`PUT /<index>/_block/<block>`

[discrete]
[[add-index-block-api-desc]]
==== {api-description-title}

You use the add index blocks API to add an index block.

include::{es-repo-dir}/indices/open-close.asciidoc[tag=closed-index]

[discrete]
[[add-index-block-api-wait-for-active-shards]]
===== Wait For active shards

include::{es-repo-dir}/indices/open-close.asciidoc[tag=wait-for-active-shards]

> [discrete]
> [[add-index-block-api-path-params]]
> ==== {api-path-parms-title}

Reviewer: This ensures the `<block>` path parameter definition renders correctly.
```suggestion
[discrete]
[role="child_attributes"]
[[add-index-block-api-path-params]]
==== {api-path-parms-title}
```

include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=index]
+
To add blocks to all indices, use `_all` or `*`. To disallow the adding
of blocks to indices with `_all` or wildcard expressions,
change the `action.destructive_requires_name` cluster setting to `true`.
You can update this setting in the `elasticsearch.yml` file
or using the <<cluster-update-settings,cluster update settings>> API.

> 

Reviewer: We should explicitly document the `<block>` path parameter.

```suggestion
`<block>`::
(Required, string)
Block type to add to the index.
+
.Valid values for `<block>`
[%collapsible%open]
====
`metadata`::
Disable metadata changes, such as closing the index.

`read`::
Disable read operations.

`read_only`::
Disable write operations and metadata changes.

`read_only_allow_delete`::
Disable write operations and metadata changes.
Document deletion is disabled.
However, index deletion is still allowed.

`write`::
Disable write operations. However, metadata changes are still allowed.
====
```

[[index-modules-blocks]]
== Index blocks

Index blocks limit the kind of operations that are available on a certain
index. The blocks come in different flavours, allowing to block write,
read, or metadata operations. The blocks can be set / removed using dynamic
index settings, or can be added using a dedicated API, which also ensures
that, once successfully returning to the user, all shards of the index are
properly accounting for the block, for example that all in-flight writes to

> an index have been completed after adding the write block.

Reviewer 2: Should we call out that this only applies to write operations, not read operations?
